Add cluster myshardid command (#2211)
redis/redis-doc: commands/cluster-nodes.md

67ed2db8d677e59ec4a4cefb06858cf2a1a89fa1 127.0.0.1:30002@31002,,shard-id=114f6674a35b84949fe567f5dfd41415ee776261 master - 0 1426238316232 2 connected 5461-10922
292f8b365bb7edb5e285caf0b7e6ddc7265d2f4f 127.0.0.1:30003@31003,,shard-id=fdb36c73e72dd027bc19811b7c219ef6e55c550e master - 0 1426238318243 3 connected 10923-16383
6ec23923021cf3ffec47632106199cb7f496ce01 127.0.0.1:30005@31005,,shard-id=114f6674a35b84949fe567f5dfd41415ee776261 slave 67ed2db8d677e59ec4a4cefb06858cf2a1a89fa1 0 1426238316232 5 connected
824fe116063bc5fcf9f4ffd895bc17aee7731ac3 127.0.0.1:30006@31006,,shard-id=fdb36c73e72dd027bc19811b7c219ef6e55c550e slave 292f8b365bb7edb5e285caf0b7e6ddc7265d2f4f 0 1426238317741 6 connected
e7d1eecce10fd6bb5eb35b9f99a514335d9ba9ca 127.0.0.1:30001@31001,,shard-id=69bc080733d1355567173199cff4a6a039a2f024 myself,master - 0 0 1 connected 0-5460
```

Each line is composed of the following fields:

```
<id> <ip:port@cport[,hostname[,auxiliary_field=value]*]> <flags> <master> <ping-sent> <pong-recv> <config-epoch> <link-state> <slot> <slot> ... <slot>
```

The meaning of each filed is the following:

1. `id`: The node ID, a 40-character globally unique string generated when a node

>    is created and never changed again (unless `CLUSTER RESET HARD` is used).

Reviewer: Leaving everything on a single line does a better job of preserving git history. It also makes it much easier to review changes since it doesn't flag the entire line as `new`.

Author: Ah, I thought we had a different preference for the documentation. Will revert next

Reviewer 2: See https://github.com/redis/redis-doc/#styling-guidelines

field called shard-id. The following is an example of output on Redis 7.2.0.

```
07c37dfeb235213a872192d90877d0cd55635b91 127.0.0.1:30004@31004,,shard-id=69bc080733d1355567173199cff4a6a039a2f024 slave e7d1eecce10fd6bb5eb35b9f99a514335d9ba9ca 0 1426238317239 4 connected
67ed2db8d677e59ec4a4cefb06858cf2a1a89fa1 127.0.0.1:30002@31002,,shard-id=114f6674a35b84949fe567f5dfd41415ee776261 master - 0 1426238316232 2 connected 5461-10922
292f8b365bb7edb5e285caf0b7e6ddc7265d2f4f 127.0.0.1:30003@31003,,shard-id=fdb36c73e72dd027bc19811b7c219ef6e55c550e master - 0 1426238318243 3 connected 10923-16383
6ec23923021cf3ffec47632106199cb7f496ce01 127.0.0.1:30005@31005,,shard-id=114f6674a35b84949fe567f5dfd41415ee776261 slave 67ed2db8d677e59ec4a4cefb06858cf2a1a89fa1 0 1426238316232 5 connected
824fe116063bc5fcf9f4ffd895bc17aee7731ac3 127.0.0.1:30006@31006,,shard-id=fdb36c73e72dd027bc19811b7c219ef6e55c550e slave 292f8b365bb7edb5e285caf0b7e6ddc7265d2f4f 0 1426238317741 6 connected
e7d1eecce10fd6bb5eb35b9f99a514335d9ba9ca 127.0.0.1:30001@31001,,shard-id=69bc080733d1355567173199cff4a6a039a2f024 myself,master - 0 0 1 connected 0-5460

> ```

Reviewer: Historically we just show the latest value and indicate what might be optional in past versions. I would just keep this latest value. We should also probably make it clear this format can change arbitrarily as is mostly textual.

Author: >We should also probably make it clear this format can change arbitrarily as is mostly textual.

This disclaimer seems a bit scary to me and I don't think it accurately captures our intent and efforts so far (to keep it as backward compatible as possible). 

Agreed with your other comments.

Reviewer: > This disclaimer seems a bit scary to me and I don't think it accurately captures our intent and efforts so far (to keep it as backward compatible as possible).

Point taken. We might want to point out the proper way to "parse" it then, so we have some rules we hold as invariants.

Author: Would the following work? 

```
4. `[,auxiliary_field=value]*`: A list of comma-separated key-value pairs that
   represent various node properties, such as `shard-id`. There is no intrinsic
   order among auxiliary fields. The auxiliary fields can appear at
   different position in the list from release to release. Both the key name and
   value can contain ASCII alphanumeric characters and the characters in
   `!#$%&()*+-.:;<>?@[]^_{|}~` only. Auxiliary fields are explained in detail in
   the section below.
```

Frankly speaking, an air-tight definition could actually hurt extensibility. Imagine that we had asked everyone to parse the second column exactly as `ip:port@cport`, we would not be able to extend it to include the new fields like `hostname`, `shard-id`, etc. I think we have struck a good balance with the callout above on the new aux fields along with the existing rules but let me know if you think other invariants should be included too.

Reviewer: I was actually going to just suggest that it is "space separated fields", the only invariant we plan on really holding is the lack of spaces. I still don't think end users should be parsing this information, and we've been pushing clients away from it.

Reviewer: I don't know if it really hurts extensibility. One of the problems we had was we never told users any way to parse, and they were guessing at it. It also was inherently inextensible before, and now at least we have a mechanism.

Reviewer: I suppose what I really still want to say is don't use this because it doesn't have a schema. Perhaps it's better to focus the efforts on the new API that provides admin information. So I'm okay leaving this, and we can mark it "deprecated" later.

Author: SGTM

Each line is composed of the following fields:

```
<id> <ip:port@cport[,hostname[,auxiliary_field=value]*]> <flags> <master> <ping-sent> <pong-recv> <config-epoch> <link-state> <slot> <slot> ... <slot>
```

The meaning of each filed is the following:

> 1. `id`: The node ID, a 40 characters random string generated when a node is created and never changed again (unless `CLUSTER RESET HARD` is used).

Reviewer 2: I think it's useful to keep this as a numbered list, because the fields are just space separated and the number is the column number each field.

> 1. `id`: The node ID, a 40 characters random string generated when a node is created and never changed again (unless `CLUSTER RESET HARD` is used).

Author: I also felt that numbered list works better so keep it for now.

2. `ip:port@cport`: The node address where clients should contact the node to run queries.

> 3. `hostname`: A human readable string that can be configured via the

Reviewer 3: ```suggestion
* `hostname`: A human readable string that can be configured via the `cluster-annouce-hostname` setting. 
The max length of the string is 256 characters, excluding the null terminator. 
The name can contain ASCII alphanumeric characters, '-', and '.' only.
```

>    `cluster-annouce-hostname` setting. The max length of the string is 256

Reviewer 3: ```suggestion

```

>    characters, excluding the null terminator. The name can contain ASCII alphanumeric characters,

Reviewer 3: ```suggestion

```

>    '-', and '.' only.

Reviewer 3: ```suggestion

```

> 4. `[,auxiliary_field=value]*`: A list of comma-separated key-value pairs that

Reviewer 3: ```suggestion
* `[,auxiliary_field=value]*`: A list of comma-separated key-value pairs that represent various node properties, such as `shard-id`. 
There is no intrinsic order among auxiliary fields. 
The auxiliary fields can appear at different positions in the list from release to release. 
Both the key name and value can contain alphanumeric characters and the characters in `!#$%&()*+-.:;<>?@[]^_{|}~` only. 
Auxiliary fields are explained below.
```

   represent various node properties, such as `shard-id`. There is no intrinsic
   order among auxiliary fields. The auxiliary fields can appear at
   different position in the list from release to release. Both the key name and
   value can contain alphanumeric characters and the characters in
   `!#$%&()*+-.:;<>?@[]^_{|}~` only. Auxiliary fields are explaned in detail in

>    the section below.

Reviewer 3: ```

> 5. `flags`: A list of comma separated flags: `myself`, `master`, `slave`, `fail?`, `fail`, `handshake`, `noaddr`, `nofailover`, `noflags`. Flags are explained in detail in the next section.

Reviewer 3: ```suggestion
* `flags`: A list of comma separated flags: `myself`, `master`, `slave`, `fail?`, `fail`, `handshake`, `noaddr`, `nofailover`, `noflags`. 
Flags are explained below.
```

> 6. `master`: If the node is a replica, and the master is known, the master node ID, otherwise the "-" character.

Reviewer 3: ```suggestion
* `master`: If the node is a replica, and the primary is known, the primary node ID, otherwise the "-" character.
```

> 7. `ping-sent`: Milliseconds unix time at which the currently active ping was sent, or zero if there are no pending pings.

Reviewer 3: ```suggestion
* `ping-sent`: UNIX time at which the currently active ping was sent, or zero if there are no pending pings, in milliseconds.
```

> 8. `pong-recv`: Milliseconds unix time the last pong was received.

Reviewer 3: ```suggestion
* `pong-recv`: UNIX time the last pong was received, in milliseconds.
```

> 9. `config-epoch`: The configuration epoch (or version) of the current node (or of the current master if the node is a replica). Each time there is a failover, a new, unique, monotonically increasing configuration epoch is created. If multiple nodes claim to serve the same hash slots, the one with higher configuration epoch wins.

Reviewer 3: ```suggestion
* `config-epoch`: The configuration epoch (or version) of the current node (or of the current primary if the node is a replica). Each time there is a failover, a new, unique, monotonically increasing configuration epoch is created. If multiple nodes claim to serve the same hash slots, the one with the higher configuration epoch wins.
```

> 10. `link-state`: The state of the link used for the node-to-node cluster bus. We use this link to communicate with the node. Can be `connected` or `disconnected`.

Reviewer 3: ```suggestion
* `link-state`: The state of the link used for the node-to-node cluster bus. Use this link to communicate with the node. Can be `connected` or `disconnected`.
```

> 11. `slot`: A hash slot number or range. Starting from argument number 9, but there may be up to 16384 entries in total (limit never reached). This is the list of hash slots served by this node. If the entry is just a number, is parsed as such. If it is a range, it is in the form `start-end`, and means that the node is responsible for all the hash slots from `start` to `end` including the start and end values.

Reviewer 3: ```suggestion
* `slot`: A hash slot number or range. Starting from argument number 9, but there may be up to 16384 entries in total (limit never reached). This is the list of hash slots served by this node. If the entry is just a number, it is parsed as such. If it is a range, it is in the form `start-end`, and means that the node is responsible for all the hash slots from `start` to `end` including the start and end values.
```

> Auxiliary fields (filed number 4):

Reviewer 3: ```suggestion
Auxiliary fields are:
```

> * `shard-id`: a 40 characters random string generated when a node is created. A

Reviewer 3: ```suggestion
* `shard-id`: a 40-random-character string generated when a node is created. A node's shard id changes only when the node joins a different shard via `cluster replicate` and there the node's shard id is updated to its primary's.
```

Reviewer: May also not be random in the future. So let's just say 40 character string.

Author: Good point. I think the real important property is "globally unique" so added it as the qualifier.

  node's shard id changes only when the node joins a different shard via
  `cluster replicate` and there the node's shard id is updated to its primary's.

> Meaning of the flags (field number 5):

Reviewer 3: ```suggestion
Flags are:
```

* `myself`: The node you are contacting.

> * `master`: Node is a master.

Reviewer 3: ```suggestion
* `master`: Node is a primary.
```
